examples/basic_query_extractor/src/main.rs

```diff
@@ -0,0 +1,141 @@
+use gotham::router::builder::{build_simple_router, DefineSingleRoute, DrawRoutes};
+use gotham::state::{State, FromState};
+
+/// `QueryParam` struct
+///
+/// It contains only a `name` field for the sake of simplicity
+/// It permits to use it as query parameter like:
+/// GET /widgets?name=t-shirt
+#[derive(Deserialize, StateData, StaticResponseExtender)]
+struct QueryParam {
+    name: String,
+}
+
+/// `Product` struct
+///
+/// It represents products that can be queried
+#[derive(Serialize)]
+struct Product {
+    name: String,
+    description: String,
+}
+
+// Generate a `Vec<Product>`
+fn generate_products() -> Vec<Product> {
+    vec![
+        Product {
+            name: "t-shirt".to_string(),
+            description: "t-shirt".to_string(),
+        },
+    ]
+}
+
+/// Returns a `Response` with the `Product` serialized to JSON
+/// If no `Product` is found, a `Response` containing a `StatusCode::NotFound`
+/// is returned.
+fn product_matcher(requested: &str, products: &Vec<Product>) -> Response {
+    match products.iter().find(|p| p.name == requested) {
+        Some(product) => {
+            Response::new()
+                .with_status(StatusCode::Ok)
+                .with_body(serde_json::to_string(product).unwrap())
+                .with_header(ContentType(mime::APPLICATION_JSON))
+        }
+        None => Response::new().with_status(StatusCode::NotFound),
+    }
+}
+
+/// Function to handle the `GET` requests coming to `/widgets/:name`
+fn get_product_handler(state: State) -> (State, Response) {
+    // Build a response
+    let res = {
+        // Extract the `QueryParam` from `state`
+        let query_param = QueryParam::borrow_from(&state);
+        let name = &query_param.name;
+        // Generate the response
+        product_matcher(&name, &generate_products())
+    };
+    (state, res)
+}
+
+/// Create a `Router`
+///
+/// /widgets?name=...             --> GET
+fn router() -> Router {
+    build_simple_router(|route| {
+        route
+            .get("/widgets")
+            .with_query_string_extractor::<QueryParam>()
+            .to(get_product_handler);
+    })
+}
+
+/// Start a server and use a `Router` to dispatch requests
+pub fn main() {
+    let addr = "127.0.0.1:7878";
+    println!("Listening for requests at http://{}", addr);
+    gotham::start(addr, router())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use gotham::test::TestServer;
+
+    #[test]
+    fn tshirt_found() {
+        let test_server = TestServer::new(router()).unwrap();
+        let response = test_server
+            .client()
+            .get("http://localhost/widgets?name=t-shirt")
+            .perform()
+            .unwrap();
+
+        assert_eq!(response.status(), StatusCode::Ok);
+
+        {
+            let content_type = response.headers().get::<ContentType>().expect(
+                "ContentType",
+            );
+            assert_eq!(content_type.0, mime::APPLICATION_JSON);
+        }
```

Comment:
I'd be happy with this test just asserting `StatusCode::Ok` and the body to minimise line count here.